src/Microsoft.DocAsCode.Common/FileAbstractLayer/LinkFileSystem.cs

```diff
@@ -0,0 +1,277 @@
+            _outputList.Add(new PathMapping(destFileName.GetPathFromWorkingFolder(), pp));
+        }
+
+        #region Help Methods
+
+        public string GetRandomEntry()
+        {
+            string name;
+            do
+            {
+                name = Path.GetRandomFileName();
+            } while (Directory.Exists(Path.Combine(OutputFolder, name)) || File.Exists(Path.Combine(OutputFolder, name)));
+            return name;
+        }
+
+        public string CreateRandomFileName() =>
+            RetryIO(() =>
+            {
+                string fileName = GetRandomEntry();
+                using (File.Create(Path.Combine(OutputFolder, fileName)))
+                {
+                    // create new zero length file.
+                }
+                return fileName;
+            });
+
+        public Tuple<string, FileStream> CreateRandomFileStream()
+        {
+            return RetryIO(() =>
+            {
+                var file = GetRandomEntry();
+                return Tuple.Create(file, File.Create(Path.Combine(OutputFolder, file)));
+            });
+        }
+
+        public static T RetryIO<T>(Func<T> func)
+        {
+            var count = 0;
+            while (true)
+            {
+                try
+                {
+                    return func();
+                }
+                catch (IOException)
+                {
+                    if (count++ >= MaxRetry)
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        public static void RetryIO(Action action)
+        {
+            var count = 0;
+            while (true)
+            {
+                try
+                {
+                    action();
+                    return;
+                }
+                catch (IOException)
+                {
+                    if (count++ >= MaxRetry)
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        #endregion
+
+        #endregion
+
+        #region Private Methods
+
+        private string FindPhysicalPath(RelativePath file)
+        {
+            var physicalPath = FindPhysicalPathNoThrow(file);
+            if (physicalPath == null)
+            {
+                throw new FileNotFoundException("File not found.", file);
+            }
+            return physicalPath;
+        }
+
+        private string FindPhysicalPathNoThrow(RelativePath file)
+        {
+            var path = file.GetPathFromWorkingFolder();
+            foreach (var m in Mappings)
+            {
+                if (m.IsFolder)
+                {
+                    var localPath = path - m.LogicalPath;
+                    if (m.AllowMoveOut || localPath.ParentDirectoryCount == 0)
+                    {
+                        var physicalPath = Path.Combine(m.PhysicalPath, localPath.ToString());
```

Comment:
PathMapping should ensure relative file path mapping to a file, anyway there is no code to ensure it.

---
In reply to: [89931868](https://github.com/dotnet/docfx/pull/926#discussion_r89931868) [](ancestors = 89931868)